source/engine/hdf5/HDF5WriterP.cpp

```diff
@@ -0,0 +1,374 @@
+
+#include <iostream> //needs to go away, this is just for demo purposes
+
+#ifdef ADIOS_HAVE_MPI
+#include "engine/hdf5/HDF5WriterP.h"
+
+#include "core/Support.h"
+#include "functions/adiosFunctions.h" //CSVToVector
+
+// supported transports
+//#include "transport/file/FD.h"      // uses POSIX
+//#include "transport/file/FP.h"      // uses C FILE*
+//#include "transport/file/FStream.h" // uses C++ fstream
+
+namespace adios {
+
+HDF5Writer::HDF5Writer(ADIOS &adios, const std::string name,
+                       const std::string accessMode, MPI_Comm mpiComm,
+                       const Method &method, const bool debugMode,
+                       const unsigned int cores)
+    : Engine(adios, "HDF5Writer", name, accessMode, mpiComm, method, debugMode,
+             cores, " HDF5Writer constructor (or call to ADIOS Open).\n"),
+      m_Buffer(accessMode, m_RankMPI, m_DebugMode) {
+  DefH5T_COMPLEX_DOUBLE =
+      H5Tcreate(H5T_COMPOUND, sizeof(ADIOS2_Complex_Double));
+  DefH5T_COMPLEX_FLOAT = H5Tcreate(H5T_COMPOUND, sizeof(ADIOS2_Complex_Float));
+  DefH5T_COMPLEX_LongDOUBLE =
+      H5Tcreate(H5T_COMPOUND, sizeof(ADIOS2_Complex_LongDouble));
+
+  // compound data type for memory
+  H5Tinsert(DefH5T_COMPLEX_DOUBLE, "double real",
+            HOFFSET(ADIOS2_Complex_Double, _re), H5T_NATIVE_DOUBLE);
+  H5Tinsert(DefH5T_COMPLEX_DOUBLE, "double img",
+            HOFFSET(ADIOS2_Complex_Double, _im), H5T_NATIVE_DOUBLE);
+  H5Tinsert(DefH5T_COMPLEX_FLOAT, "float real",
+            HOFFSET(ADIOS2_Complex_Float, _re), H5T_NATIVE_FLOAT);
+  H5Tinsert(DefH5T_COMPLEX_FLOAT, "float img",
+            HOFFSET(ADIOS2_Complex_Float, _im), H5T_NATIVE_FLOAT);
+  H5Tinsert(DefH5T_COMPLEX_LongDOUBLE, "ldouble real",
+            HOFFSET(ADIOS2_Complex_LongDouble, _re), H5T_NATIVE_LDOUBLE);
+  H5Tinsert(DefH5T_COMPLEX_LongDOUBLE, "ldouble img",
+            HOFFSET(ADIOS2_Complex_LongDouble, _im), H5T_NATIVE_LDOUBLE);
+
+  // data type for file
+  int ldsize = H5Tget_size(H5T_NATIVE_DOUBLE);
+  DefH5T_filetype_COMPLEX_DOUBLE = H5Tcreate(H5T_COMPOUND, ldsize + ldsize);
+  H5Tinsert(DefH5T_filetype_COMPLEX_DOUBLE, "double real", 0, H5T_IEEE_F64BE);
+  H5Tinsert(DefH5T_filetype_COMPLEX_DOUBLE, "double img", ldsize,
+            H5T_IEEE_F64BE);
+
+  ldsize = H5Tget_size(H5T_NATIVE_FLOAT);
+  DefH5T_filetype_COMPLEX_FLOAT = H5Tcreate(H5T_COMPOUND, ldsize + ldsize);
+  H5Tinsert(DefH5T_filetype_COMPLEX_FLOAT, "float real", 0, H5T_IEEE_F32BE);
+  H5Tinsert(DefH5T_filetype_COMPLEX_FLOAT, "float img", ldsize, H5T_IEEE_F32BE);
+
+  ldsize = H5Tget_size(
+      H5T_NATIVE_LDOUBLE); // note that sizeof(H5T_NATIVE_LDOUBLE) returned 4!!
+  DefH5T_filetype_COMPLEX_LongDOUBLE = H5Tcreate(H5T_COMPOUND, ldsize + ldsize);
+  H5Tinsert(DefH5T_filetype_COMPLEX_LongDOUBLE, "ldouble real", 0,
+            H5T_NATIVE_LDOUBLE);
+  H5Tinsert(DefH5T_filetype_COMPLEX_LongDOUBLE, "ldouble img", ldsize,
+            H5T_NATIVE_LDOUBLE);
+
+  Init();
+}
+
+HDF5Writer::~HDF5Writer() {}
```

Comment:
Remove empty implementation and use `virtual ~HDFWriter() = default;` in header.